Issue 2281: Tier-1 Repair Tool

RaulGracia opened the pull request:
**Change log description**  
Added a new tool in Admin CLI that allows to "edit" Operations in a DurableDataLog (see https://github.com/pravega/pravega/wiki/PDP-54-(Tier-1-Repair-Tool)). 

**Purpose of the change**  
Fixes #2281.

**What the code does**  
This PR adds a new commands to the Admin CLI, namely `DurableDataLogRepairCommand`. It allows to "edit" a `DurableDataLog` that contains one or more damaged `Operations` on it. These are the most important changes in this PR:
- We cannot actually "edit" a `DurableDataLog`, as it is an append-only data structure. To achieve this notion of editing, `DurableDataLogRepairCommand` copies, processes edits and overwrites the originally damaged `DurableDataLog` in separate steps.
  - *1st Step*: The command creates a Backup Log from the Original Log. This Backup Log is stored with a special log id (`BACKUP_LOG=Integer.MAX_VALUE-1`).
    - Note that in the process of performing the Backup Log, it has been decided to discard repeated `Operation`s (i.e., attempts to perform new `Operation` writes with old sequence numbers). Such a basic filtering would fix problems of logs with duplicate entries without any further action.
  - *2nd Step*: Once created the Backup Log, the `DurableDataLogRepairCommand` asks the user to introduce the edits wanted on the Original Log. This basically implies to add zero or more `LogEditOperation`s to the list of edits to be applies to the Original Log.
    - An  `LogEditOperation` can be `DELETE_OPERATION` (i.e., the Original Log will not contain the `Operation`s defined within the initial (inclusive) and final (exclusive) sequence numbers of it); `ADD_OPERATION` (i.e., the user can add one or more `Operation`s at the point of the Original Log specified by the initial sequence number of it); and `REPLACE_OPERATION` (i.e., the user can replace one `Operation` by another as specified in the sequence number of it).
    - Some `Operation`s require a binary payload (`MetadataCheckpointOperation`, `StorageMetadataCheckpointOperation`, and `StreamSegmentAppendOperation`). At the moment, we offer 2 ways of creating such a payload: i) `zeros`, so the tool will create binary contents of specified length filled with zeros; ii) `file`, the tool can read a file with a provided binary payload to be used in that `Operation`.
    - The final step of this process is to read the data from the Backup Log and store a new log, called Repair Log, by applying the `LogEditOperation`s to the right `Operation`s the Backup log. This means that the Repair Log will contain the contents of the Backup Log, plus adding, replacing and/or deleting `Operation`s as specified by the user. The Repair Log is created on a special log id as well (`REPAIR_LOG=Integer.MAX_VALUE`).
    - Note that the `REPAIR_LOG` starts sequence numbers from 1 and assigns a new sequence number to each new `Operation` added from that point onwards (even though the `Operation`s had a higher sequence number in the Original Log). This also required to add a new method called `resetSequenceNumber` in `Operation` class.
  - *3rd Step*:  The final step consists of overwriting the Original Log metadata with that of Repair Log. The objective is that, when the Original Log is re-enabled, it will read the ledgers corresponding to the Repair Log without doing any further change in Pravega. If the edits actually solve the problem, the Segment Container will be able to recover.
    - As part of the Segment Container recovery process, the `LogReader` checks that the ledgers read actually belong to the log id of the log at hand. This PR adds an exception to this check, as ledgers from `REPAIR_LOG` should be also allowed during the repair process.

- When re-executing the command, the tool will check if there is any BACKUP_LOG. If so, the user can decide to remove it (in case it is not needed anymore), or to re-use it. In the latter case, this allows us to perform multiple retries to repair the Original Log in the case the first attempt was not successful.
- Most of the logic in the command expects a `DurableDataLog` in an attempt to make it as generic as possible for alternative Tier-1 implementations. However, there is functionality required that is only available today in `DurableLogWrapper` class, which is part of Bookkeeper Tier-1 implementation. In the future, we can re-think the interfaces that Tier-1 implementations need to offer for admin purposes to make this command completely Tier-1 agnostic.

**How to verify it**  
Added new tests to verify most of the functionality.
(Need to exercise the new command on a real cluster)

andreipaduroiu commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+    @VisibleForTesting
+    Operation createUserDefinedOperation() {
+        Operation result;
+        final String operations = "[DeleteSegmentOperation|MergeSegmentOperation|MetadataCheckpointOperation|" +
```
We'd want to thoroughly document these options and the entire syntax/workflow for each of them. I can see an admin that isn't comfortable with the code would be reluctant to making any changes here.

RaulGracia replied:
I think that this command (and perhaps the other DR commands) require a separate documentation page describing in depth their operation, workflow and so on. Will create it and complete the part related to this command.

RaulGracia replied:
Committed documentation for that.

andreipaduroiu commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+                        output("Wrong option. Please, select one of the following options: [zero|file]");
+                }
+            } catch (Exception ex) {
+                outputError("Some problem has happened.");
```
I can see some unexpected error getting the user into some infinite loop with no way to get out. Maybe we should provide an option to abort?

RaulGracia replied:
Added possibility to abort this process.

andreipaduroiu commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+    @VisibleForTesting
+    SegmentProperties createSegmentProperties() {
+        String segmentName = getStringUserInput("Introduce the name of the Segment: ");
+        long offset = getLongUserInput("Introduce the offset of the Segment: ");
```
I find the "Introduce" to be a non-standard way to request something from the command line. I've seen "Input" used before.

RaulGracia replied:
Fixed.

andreipaduroiu commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+            output("Creating an AttributeUpdateCollection for this operation.");
+            try {
+                AttributeId attributeId = AttributeId.fromUUID(UUID.fromString(getStringUserInput("Introduce UUID for this AttributeUpdate: ")));
+                AttributeUpdateType type = AttributeUpdateType.get((byte) getIntUserInput("Introduce AttributeUpdateType for this AttributeUpdate" +
```
FYI once serialized in the log, the update type and comparison value REALLY do not matter anymore. They are pre-validated initially when we get the operation but later on the value has already been precomputed and accepted. So I think we can simplify this by removing "AttributeUpdateType" and "Comparison value". Just use `Replace`.

RaulGracia replied:
Yes, but if someone looks at an `AttributeUpdate` to understand what it is, he/she may be confused about just adding some fields and not others. Even though they may not be necessary, the current approach may be more directly understandable from a non-expert person in the Segment Store.

andreipaduroiu commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+        while (!finishInputCommands) {
+            output("Creating an AttributeUpdateCollection for this operation.");
+            try {
+                AttributeId attributeId = AttributeId.fromUUID(UUID.fromString(getStringUserInput("Introduce UUID for this AttributeUpdate: ")));
```
Not all attributes are UUIDs. The `FixedKeyLengthTableSegment` (find the class that starts with that prefix) uses different Attribute ID lengths. 

Maybe we can improve on this later, but since those segments are also part of the container, we should figure out a way to fix the recovery process if it's stuck on one of them.

RaulGracia replied:
Right, if you believe that this is an important limitation, I can think on how to do this. If not, we can think about this on a further iteration of the tool.

andreipaduroiu reviewed (commented):
This is great work. I am not done yet - I will do another pass tomorrow, but feel free to look at the comments I made so far.

abhinb commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+        val originalDataLog = dataLogFactory.createDebugLogWrapper(containerId);
+
+        // Disable the Original Log, if not disabled already.
+        output("Original DurableLog is enabled. You are about to disable it.");
```
Is it possible to tell the user that the log is disabled and we can proceed further instead of trying to disable an already disabled log?

abhinb commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
does it make sense in a further iteration/version of this tool to provide an option to scan for any logs that could be corrupted and provide an option to repair?

RaulGracia replied:
If you men that, during the Backup phase, we could hint the admin, I think that it is a good idea. Some hints may be simple, but I image that others may not be as simple. Good idea.

abhinb commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+        val originalDataLog = dataLogFactory.createDebugLogWrapper(containerId);
+
+        // Disable the Original Log, if not disabled already.
+        output("Original DurableLog is enabled. You are about to disable it.");
```
Is it possible to tell the user that the log is disabled and we can proceed further instead of trying to disable an already disabled log?

RaulGracia replied:
I think that the current version is more aligned with what you are asking here.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java:
```diff
@@ -166,8 +172,24 @@ protected CuratorFramework createZKClient() {
+     */
     protected void output(String messageTemplate, Object... args) {
-        this.out.println(String.format(messageTemplate, args));
+        this.out.printf(messageTemplate + "%n", args);
```
Right, missed that one.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DataRecoveryCommand.java:
```diff
@@ -17,19 +17,60 @@
+    protected static final String COMPONENT = "storage";
+
+    protected static final int REPAIR_LOG_ID = Integer.MAX_VALUE;
+    protected static final int BACKUP_LOG_ID = Integer.MAX_VALUE - 1;
```
Yes, you are right and I'm aware of that. However, given the nature of this tool, having only one repair at a time does not seem a critical limitation (having multiple admins doing dangerous operation at the same time on the same cluster seems kind of non-advisable). I have specified this in the "Limitations" section of the PDP: https://github.com/pravega/pravega/wiki/PDP-54-(Tier-1-Repair-Tool)#limitations

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DataRecoveryCommand.java:
```diff
@@ -17,19 +17,60 @@
+            .builder()
+            .with(DurableLogConfig.CHECKPOINT_MIN_COMMIT_COUNT, 10000)
+            .with(DurableLogConfig.CHECKPOINT_COMMIT_COUNT, 50000)
+            .with(DurableLogConfig.CHECKPOINT_TOTAL_COMMIT_LENGTH, 1024 * 1024 * 1024L)
```
I actually just moved this code from `DurableLogRecoveryCommand` to its parent class, as I used some of these in the early stages of development of this PR. Now, I revisited that, and I reverted this change as I'm not using it, so all this is again part of `DurableLogRecoveryCommand`. If this configuration is not correct and we have found it now, maybe we could create a separate issue to fix it.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DataRecoveryCommand.java:
```diff
@@ -17,19 +17,60 @@
+
+    protected static final ContainerConfig CONTAINER_CONFIG = ContainerConfig
+            .builder()
+            .with(ContainerConfig.SEGMENT_METADATA_EXPIRATION_SECONDS, 10 * 60)
```
I actually just moved this code from `DurableLogRecoveryCommand` to its parent class, as I used some of these in the early stages of development of this PR. Now, I revisited that, and I reverted this change as I'm not using it, so all this is again part of `DurableLogRecoveryCommand`. If this configuration is not correct and we have found it now, maybe we could create a separate issue to fix it.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DataRecoveryCommand.java:
```diff
@@ -53,25 +94,42 @@ StorageFactory createStorageFactory(ScheduledExecutorService executorService) {
+        public final CacheManager cacheManager;
+
+        Context(ScheduledExecutorService scheduledExecutorService) {
+            this.cacheStorage = new DirectMemoryCache(Integer.MAX_VALUE / 5);
```
I actually just moved this code from `DurableLogRecoveryCommand` to its parent class, as I used some of these in the early stages of development of this PR. Now, I revisited that, and I reverted this change as I'm not using it, so all this is again part of `DurableLogRecoveryCommand`. If this configuration is not correct and we have found it now, maybe we could create a separate issue to fix it.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DataRecoveryCommand.java:
```diff
@@ -53,25 +94,42 @@ StorageFactory createStorageFactory(ScheduledExecutorService executorService) {
+        public void close() {
+            this.readIndexFactory.close();
+            this.cacheManager.close();
+            this.cacheStorage.close();
```
I actually just moved this code from `DurableLogRecoveryCommand` to its parent class, as I used some of these in the early stages of development of this PR. Now, I revisited that, and I reverted this change as I'm not using it, so all this is again part of `DurableLogRecoveryCommand`. If this configuration is not correct and we have found it now, maybe we could create a separate issue to fix it.

RaulGracia commented on cli/admin/src/test/java/io/pravega/cli/admin/segmentstore/AbstractSegmentStoreCommandsTest.java:
```diff
@@ -230,7 +230,7 @@ public void testUpdateSegmentAttributeCommand() throws Exception {
     public void testFlushToStorageCommandAllCase() throws Exception {
         String commandResult = TestUtils.executeCommand("container flush-to-storage all localhost", STATE.get());
-        for (int id = 1; id < CONTAINER_COUNT; id++) {
+        for (int id = 0; id < CONTAINER_COUNT; id++) {
```
Yes, just found that this test was not testing anything :)

RaulGracia commented on segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java:
```diff
@@ -48,6 +48,10 @@
 
+    // LogReader can only read from ledgers tagged with the same log id. For special log repair operations, we allow the
+    // LogReader to read from a special log id that contains admin-provided changes to repair the original lod data.
+    private static final int REPAIR_LOG_ID = Integer.MAX_VALUE;
```
I have done some refactor of the APIs (will describe this in the PR for clarity). But with refactor, the CLI is abstracted of the actual Id of the log, and these reserved log ids are defined once per-implementation.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+/**
+ * This command provides an administrator with the basic primitives to manipulate a DurableLog with damaged entries.
+ * The workflow of this command is as follows:
+ * 1. Disable the original DurableLog (if not done yet).
```
Done, now the command exists if the Original log is enabled.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+            originalDataLog.disable();
+        } catch (DurableDataLogException ex) {
+            outputError("Problem while disabling the Original Log. It may be disabled already.");
+            ex.printStackTrace();
```
This is not needed anymore due to the previous change.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+        if (createNewBackupLog) {
+            createBackupLog(dataLogFactory, containerId, originalDataLog);
+        }
+        int backupLogRadOperations = validateBackupLog(dataLogFactory, containerId, originalDataLog);
```
Fixed, thanks.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+
+        output("Original DurableLog has been backed up correctly. Ready to apply admin-provided changes to the Original Log.");
+        if (!confirmContinue()) {
+            output("Not editing original DurableLog this time.");
```
Done.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+        protected void writeAndConfirm(Operation operation) throws IOException {
+            sequenceNumber.compareAndSet(Long.MIN_VALUE, operation.getSequenceNumber());
+            // We only consider writing operations with sequence number higher than the expected one.
+            if (operation.getSequenceNumber() >= sequenceNumber.get()) {
```
Correct.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+
+    @Override
+    public void execute() throws Exception {
+        ensureArgCount(1);
```
Now, the command can be only executed when the Original Log is disabled. This guarantees that the system is not editing it. On the other hand, once the log is disabled, we cannot control if more than one admin is running the command at the same time, which would be a bad scenario. It is a limitation, as noted here https://github.com/pravega/pravega/wiki/PDP-54-(Tier-1-Repair-Tool)#limitations, but an affordable one: these repair operations may be quite complex, so unless it is by accident, I don't think that there would be many cases in which multiple admins start executing this command in parallel.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+            readDurableDataLogWithCustomCallback(logEditState, BACKUP_LOG_ID, backupDataLog);
+            assert !logEditState.isFailed;
+        } catch (Exception ex) {
+            outputError("There have been errors while creating the edited version of the DurableLog.");
```
Done.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,914 @@
+     */
+    @VisibleForTesting
+    void checkBackupLogAssertions(long beforeCommitCalls, long commitSuccessCalls, long operationsReadFromOriginalLog, boolean isFailed) {
+        assert beforeCommitCalls == commitSuccessCalls : "BackupLogProcessor has different number of processed (" + beforeCommitCalls +
```
Done.

RaulGracia commented on segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentInformation.java:
```diff
@@ -21,11 +21,13 @@
 /**
  * General Stream Segment Information.
  */
+@EqualsAndHashCode
```
I use this in the tests and I can properly compare 2 objects of type `StreamSegmentInformation`. I looked at the constructor of this object and it sets `lastModified` with a new `ImmutableDate` object only if none is provided via the builder, which seems reasonable. I think that is fine at the moment not excluding any field from `@EqualsAndHashCode`, as one can set all these fields and the equality works well if the values of 2 different objects match.

andreipaduroiu commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+imagine that by some reason, the Durable Log contains a duplicate Operation written (e.g., a network glitch
+has uncovered a bug in the write path). If the Segment Container attempts to recover when there is a duplicate
+Operation in the Durable Log, the recovery process will halt throwing a `DataCorruptionException` and the DurableLog will be disabled.
+This tool can help "editing" the original Durable Log to _remove_ the duplicate Operation and recover the
```
editing -> edit

RaulGracia replied:
Fixed.

andreipaduroiu commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+that gap in the original Durable Log, or even _replacing_ a malformed Operation by another one. 
+
+### Non-Targeted Scenarios 
+This tool assumes that the metadata of the Durable Log is correct and just
```
"Just don't say 'just'". It is a word that provides absolutely no meaning.

RaulGracia replied:
Fixed.

andreipaduroiu commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+iii) quit. Note that the option of re-using the existing Backup Log gives an administrator the opportunity 
+to re-try repair the Original Log, in case the previous attempt does not solve the problem and the Segment
+Container is still unable to recover.
+- Once the Backup Log is created, the tool will ask the administrator to input edits on to be done to the
```
remove `on` ?

RaulGracia replied:
Fixed.

andreipaduroiu commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+- Once the Backup Log is created, the tool will ask the administrator to input edits on to be done to the
+Original Log contents. Note that the creation of log edits may be non-trivial, so we provide more details in
+the next section.
+- With the desired edits to the Backup Log in place, the tool read the Backup Log and apply them to create
```
read -> reads

RaulGracia replied:
Fixed. In this case, "them" refers to the previous sentence "desired edits to the Backup Log".

andreipaduroiu commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+
+### Usage Instructions
+
+The Tier-1 Repair Tool is, in general, simple to use: the command just expects the id of the Segment Container
```
remove `just`.

RaulGracia replied:
I have "just" deleted all the occurrences of "just" :)

andreipaduroiu commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+
+The previous Log Edit Operation will give us the flexibility to edit a Durable Log in any way. However, there
+is one more caveat to it: Adding and replacing Operations in a Durable Log require to create them somehow. The
+tool also helps the administrator to create new Operations, and thus it will aks the admin for all the required
```
aks -> ask

RaulGracia replied:
Fixed.

RaulGracia commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+To modify a Durable Log content, the tool allows the user to create Log Edit Operations. Internally, a Log Edit 
+Operation has an _initial sequence id_, a _final sequence id_, a _type_, and _new Operation_. Log Edit Operations
+will instruct the tool how to actually do the changes on the damaged Durable Log. There are three types of edits:
+- **Delete Edit Operation**: A Log Edit Operation that will instruct the tool to do not write one or more Operations
```
Simplified that sentence.

RaulGracia commented on documentation/src/docs/admin-guide/data-recovery-tools.md:
```diff
@@ -0,0 +1,134 @@
+specified by the initial sequence id field.
+
+The previous Log Edit Operation will give us the flexibility to edit a Durable Log in any way. However, there
+is one more caveat to it: Adding and replacing Operations in a Durable Log require to create them somehow. The
```
Fixed.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DataRecoveryCommand.java:
```diff
@@ -20,16 +20,14 @@
-@Slf4j
 public abstract class DataRecoveryCommand extends AdminCommand {
-    protected final static String COMPONENT = "storage";
+    protected final static String COMPONENT = "data-recovery";
```
Right, changed that in the README file.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,918 @@
+            Preconditions.checkState(!logEditState.isFailed);
+        } catch (Exception ex) {
+            outputError("There have been errors while creating the edited version of the DurableLog.");
+            ex.printStackTrace();
```
Done.

RaulGracia commented on cli/admin/src/main/java/io/pravega/cli/admin/dataRecovery/DurableDataLogRepairCommand.java:
```diff
@@ -0,0 +1,918 @@
+
+        @Override
+        public int hashCode() {
+            int hash = 7;
```
Done.

andreipaduroiu commented:
@RaulGracia Let me know when you want this merged in. I understand you're still doing testing to validate it works as expected.

RaulGracia commented:
I have done another round of testing on a real cluster and at this point, things are working well in the cases tested. These are the results of the test:

1) After disabled container 23, run the new command:
```
> data-recovery durableLog-repair 23
...
Backing up: StreamSegmentAppendOperation: SequenceNumber = 690, SegmentId = 2, Offset = 10162, Length = 134, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 691, SegmentId = 11, Offset = 77833, Length = 699, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 692, SegmentId = 11, Offset = 78532, Length = 70, Attributes = 0
Backing up: UpdateAttributesOperation: SequenceNumber = 693, SegmentId = 11, Attributes = 4
Backing up: StreamSegmentAppendOperation: SequenceNumber = 694, SegmentId = 11, Offset = 78602, Length = 230, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 695, SegmentId = 11, Offset = 78832, Length = 114, Attributes = 0
Backing up: UpdateAttributesOperation: SequenceNumber = 696, SegmentId = 11, Attributes = 3
Backing up: UpdateAttributesOperation: SequenceNumber = 697, SegmentId = 11, Attributes = 5
Backing up: UpdateAttributesOperation: SequenceNumber = 698, SegmentId = 11, Attributes = 2
Number of operations read from DurableLog: 62
...
Number of operations read from DurableLog: 62
Original DurableLog operations read: 62, Backup DurableLog operations read: 62
```
The backup phase has been done correctly (62 operations).

2) Done some edits in the log: i) two Add Edit Operations at sequence number 696 (both DeleteSegmentOperation), ii) one Delete Edit Operation from 697 (inclusive) to 698 (exclusive), and iii) one Replace Edit Operation at 698 (another DeleteSegmentOperation). The user input can be seen here:

```
Select edit action on DurableLog: [delete|add|replace]
add
At which Operation sequence number would you like to add new Operations?
696
Type one of the following Operations to instantiate: [DeleteSegmentOperation|MergeSegmentOperation|MetadataCheckpointOperation|StorageMetadataCheckpointOperation|StreamSegmentAppendOperation|StreamSegmentMapOperation|StreamSegmentSealOperation|StreamSegmentTruncateOperation|UpdateAttributesOperation]
DeleteSegmentOperation
Input Segment Id for DeleteSegmentOperation:
1111
Input Segment Offset for StreamSegmentSealOperation:
1111
You can add more Operations at this sequence number or not.
Do you want to continue?[yes|no]
yes
Type one of the following Operations to instantiate: [DeleteSegmentOperation|MergeSegmentOperation|MetadataCheckpointOperation|StorageMetadataCheckpointOperation|StreamSegmentAppendOperation|StreamSegmentMapOperation|StreamSegmentSealOperation|StreamSegmentTruncateOperation|UpdateAttributesOperation]
DeleteSegmentOperation
Input Segment Id for DeleteSegmentOperation:
2222
Input Segment Offset for StreamSegmentSealOperation:
2222
You can add more Operations at this sequence number or not.
Do you want to continue?[yes|no]
no
You can continue adding edits to the original DurableLog.
Do you want to continue?[yes|no]
yes
Select edit action on DurableLog: [delete|add|replace]
delete
Initial operation id to delete? (inclusive)
697
Final operation id to delete? (exclusive)
698
You can continue adding edits to the original DurableLog.
Do you want to continue?[yes|no]
yes
Select edit action on DurableLog: [delete|add|replace]
replace
What Operation sequence number would you like to replace?
698
Type one of the following Operations to instantiate: [DeleteSegmentOperation|MergeSegmentOperation|MetadataCheckpointOperation|StorageMetadataCheckpointOperation|StreamSegmentAppendOperation|StreamSegmentMapOperation|StreamSegmentSealOperation|StreamSegmentTruncateOperation|UpdateAttributesOperation]
DeleteSegmentOperation
Input Segment Id for DeleteSegmentOperation:
3333
Input Segment Offset for StreamSegmentSealOperation:
3333
You can continue adding edits to the original DurableLog.
Do you want to continue?[yes|no]
no
The following edits will be used to edit the Original Log:
DurableDataLogRepairCommand.LogEditOperation(type=ADD_OPERATION, initialOperationId=696, finalOperationId=696, newOperation=DeleteSegmentOperation: SequenceNumber = -9223372036854775808, SegmentId = 1111, SegmentId = 1111)
DurableDataLogRepairCommand.LogEditOperation(type=ADD_OPERATION, initialOperationId=696, finalOperationId=696, newOperation=DeleteSegmentOperation: SequenceNumber = -9223372036854775808, SegmentId = 2222, SegmentId = 2222)
DurableDataLogRepairCommand.LogEditOperation(type=DELETE_OPERATION, initialOperationId=697, finalOperationId=698, newOperation=null)
DurableDataLogRepairCommand.LogEditOperation(type=REPLACE_OPERATION, initialOperationId=698, finalOperationId=698, newOperation=DeleteSegmentOperation: SequenceNumber = -9223372036854775808, SegmentId = 3333, SegmentId = 3333)
```
3) And this is the resulting Repair Log (interesting part only):
```
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 57, SegmentId = 11, Attributes = 4
Repair Log Operations: StreamSegmentAppendOperation: SequenceNumber = 58, SegmentId = 11, Offset = 78602, Length = 230, Attributes = 0
Repair Log Operations: StreamSegmentAppendOperation: SequenceNumber = 59, SegmentId = 11, Offset = 78832, Length = 114, Attributes = 0
Repair Log Operations: DeleteSegmentOperation: SequenceNumber = 60, SegmentId = 1111, SegmentId = 1111
Repair Log Operations: DeleteSegmentOperation: SequenceNumber = 61, SegmentId = 2222, SegmentId = 2222
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 62, SegmentId = 11, Attributes = 3
Repair Log Operations: DeleteSegmentOperation: SequenceNumber = 63, SegmentId = 3333, SegmentId = 3333
Number of operations read from DurableLog: 63
Edited DurableLog Operations read: 63
```
Visibly, we have 2 `DeleteSegmentOperation` at sequence numbers 60 and 61 (recall that in the Repair Log sequence numbers start from 1), which go before the existing Operation `UpdateAttributesOperation: SequenceNumber = 62, SegmentId = 11, Attributes = 3`. Also, the Operation `UpdateAttributesOperation: SequenceNumber = 697, SegmentId = 11, Attributes = 5` at former Sequence number 697 is not there anymore. Finally, the last Operation `UpdateAttributesOperation: SequenceNumber = 698, SegmentId = 11, Attributes = 2` in the Original Log has been replaced by `DeleteSegmentOperation: SequenceNumber = 63, SegmentId = 3333, SegmentId = 3333`, as we instructed. This shows the capabilities of the tool on a real cluster to modify logs.

4) These are the details of the BookkeeperLog after the repair happened. The increased "version" denotes that we have updated the metadata more than other logs:
```
> bk details 23
{
  "key": "log_summary",
  "value": {
    "logId": 23,
    "epoch": 1,
    "version": 5,
    "enabled": false,
    "ledgers": 1,
    "truncation": "Sequence = 0, LedgerId = 0, EntryId = 0"
  }
}
{
  "key": "ledger_details",
  "value": {
    "ledger": 985,
    "seq": 1,
    "status": "Unknown",
    "lac": 62,
    "length": 13875,
    "numBookies": 3,
    "numFragments": 1,
    "ensembleSize": 3,
    "writeQuorumSize": 3,
    "ackQuorumSize": 3,
    "ensembles": "0: [bookkeeper-bookie-0.bookkeeper-bookie-headless.default.svc.cluster.local:3181,bookkeeper-bookie-1.bookkeeper-bookie-headless.default.svc.cluster.local:3181,bookkeeper-bookie-2.bookkeeper-bookie-headless.default.svc.cluster.local:3181]"
  }
}
```

5) Now, imagine that we did a mistake and the Repair log does not work as expected. The workflow of the tool allows us to repeat the same process several times, as long as we do not delete the Backup Log. Here, we execute again the repair process on the same log:
```
>  data-recovery durableLog-repair 23
We found data in the Backup log, probably from a previous repair operation (or someone else running the same command at the same time). You have three options: 1) Delete existing Backup Log and start a new repair process, 2) Keep existing Backup Log and re-use it for the current repair (i.e., skip creating a new Backup Log), 3) Quit.
Select an option: [1|2|3]
2
...
Reading: UpdateAttributesOperation: SequenceNumber = 693, SegmentId = 11, Attributes = 4
Reading: StreamSegmentAppendOperation: SequenceNumber = 694, SegmentId = 11, Offset = 78602, Length = 230, Attributes = 0
Reading: StreamSegmentAppendOperation: SequenceNumber = 695, SegmentId = 11, Offset = 78832, Length = 114, Attributes = 0
Reading: UpdateAttributesOperation: SequenceNumber = 696, SegmentId = 11, Attributes = 3
Reading: UpdateAttributesOperation: SequenceNumber = 697, SegmentId = 11, Attributes = 5
Reading: UpdateAttributesOperation: SequenceNumber = 698, SegmentId = 11, Attributes = 2
Number of operations read from DurableLog: 62
Original DurableLog operations read: 63, Backup DurableLog operations read: 62
```
As you can see, we have read from the Original Log 63 Operations, whereas the Backup Log still contains 62. This is because now, the Original Log contains the edited version of the log from the previous execution of the tool, in which we added 2 operations and deleted 1 (so it has 1 more operations than the Backup Log).

6) Now, we are going to delete 10 Operations. Also, we are going to introduce wrong inputs in purpose: first, we are going to create an Add Edit Operation with a sequence number lower than the first Delete Edit Operation created (the user should reason and input the edits to the Backup Log in sequence). Second, we attempt to replace an Operation within the range of the sequence numbers encompassed by the Delete Edit Operation:
```
Select edit action on DurableLog: [delete|add|replace]
delete
Initial operation id to delete? (inclusive)
680
Final operation id to delete? (exclusive)
691
You can continue adding edits to the original DurableLog.
Do you want to continue?[yes|no]
yes
Select edit action on DurableLog: [delete|add|replace]
add
At which Operation sequence number would you like to add new Operations?
500
Type one of the following Operations to instantiate: [DeleteSegmentOperation|MergeSegmentOperation|MetadataCheckpointOperation|StorageMetadataCheckpointOperation|StreamSegmentAppendOperation|StreamSegmentMapOperation|StreamSegmentSealOperation|StreamSegmentTruncateOperation|UpdateAttributesOperation]
DeleteSegmentOperation
Input Segment Id for DeleteSegmentOperation:
1234
Input Segment Offset for StreamSegmentSealOperation:
1234
You can add more Operations at this sequence number or not.
Do you want to continue?[yes|no]
no
You can continue adding edits to the original DurableLog.
Do you want to continue?[yes|no]
yes
Select edit action on DurableLog: [delete|add|replace]
replace
What Operation sequence number would you like to replace?
685
Type one of the following Operations to instantiate: [DeleteSegmentOperation|MergeSegmentOperation|MetadataCheckpointOperation|StorageMetadataCheckpointOperation|StreamSegmentAppendOperation|StreamSegmentMapOperation|StreamSegmentSealOperation|StreamSegmentTruncateOperation|UpdateAttributesOperation]
DeleteSegmentOperation
Input Segment Id for DeleteSegmentOperation:
9999
Input Segment Offset for StreamSegmentSealOperation:
9999
You can continue adding edits to the original DurableLog.
Do you want to continue?[yes|no]
no
The following edits will be used to edit the Original Log:
DurableDataLogRepairCommand.LogEditOperation(type=DELETE_OPERATION, initialOperationId=680, finalOperationId=691, newOperation=null)
Original DurableLog has been backed up correctly. Ready to apply admin-provided changes to the Original Log.
Do you want to continue?[yes|no]
yes
```
Visibly, the tool only allows the Delete Edit Operation, as it is the only one that is formally correct.

7) Next, we observe that the deletion of 10 Operations has completed correctly:

```
(Backup Log)
Backing up: UpdateAttributesOperation: SequenceNumber = 678, SegmentId = 11, Attributes = 5
Backing up: StreamSegmentAppendOperation: SequenceNumber = 679, SegmentId = 2, Offset = 10028, Length = 134, Attributes = 0
Backing up: UpdateAttributesOperation: SequenceNumber = 680, SegmentId = 11, Attributes = 2
Backing up: StreamSegmentAppendOperation: SequenceNumber = 681, SegmentId = 11, Offset = 76462, Length = 751, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 682, SegmentId = 11, Offset = 77213, Length = 0, Attributes = 0
Backing up: UpdateAttributesOperation: SequenceNumber = 683, SegmentId = 11, Attributes = 6
Backing up: StreamSegmentTruncateOperation: SequenceNumber = 684, SegmentId = 2, Offset = 10028
Backing up: StreamSegmentAppendOperation: SequenceNumber = 685, SegmentId = 11, Offset = 77213, Length = 550, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 686, SegmentId = 11, Offset = 77763, Length = 70, Attributes = 0
Backing up: UpdateAttributesOperation: SequenceNumber = 687, SegmentId = 11, Attributes = 3
Backing up: UpdateAttributesOperation: SequenceNumber = 688, SegmentId = 11, Attributes = 2
Backing up: StreamSegmentTruncateOperation: SequenceNumber = 689, SegmentId = 2, Offset = 10162
Backing up: StreamSegmentAppendOperation: SequenceNumber = 690, SegmentId = 2, Offset = 10162, Length = 134, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 691, SegmentId = 11, Offset = 77833, Length = 699, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 692, SegmentId = 11, Offset = 78532, Length = 70, Attributes = 0
Backing up: UpdateAttributesOperation: SequenceNumber = 693, SegmentId = 11, Attributes = 4
Backing up: StreamSegmentAppendOperation: SequenceNumber = 694, SegmentId = 11, Offset = 78602, Length = 230, Attributes = 0
Backing up: StreamSegmentAppendOperation: SequenceNumber = 695, SegmentId = 11, Offset = 78832, Length = 114, Attributes = 0
Backing up: UpdateAttributesOperation: SequenceNumber = 696, SegmentId = 11, Attributes = 3
Backing up: UpdateAttributesOperation: SequenceNumber = 697, SegmentId = 11, Attributes = 5
Backing up: UpdateAttributesOperation: SequenceNumber = 698, SegmentId = 11, Attributes = 2
Number of operations read from DurableLog: 62

(Repair Log)
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 41, SegmentId = 11, Attributes = 2
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 42, SegmentId = 11, Attributes = 5
Repair Log Operations: StreamSegmentAppendOperation: SequenceNumber = 43, SegmentId = 2, Offset = 10028, Length = 134, Attributes = 0
Repair Log Operations: StreamSegmentAppendOperation: SequenceNumber = 44, SegmentId = 11, Offset = 77833, Length = 699, Attributes = 0
Repair Log Operations: StreamSegmentAppendOperation: SequenceNumber = 45, SegmentId = 11, Offset = 78532, Length = 70, Attributes = 0
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 46, SegmentId = 11, Attributes = 4
Repair Log Operations: StreamSegmentAppendOperation: SequenceNumber = 47, SegmentId = 11, Offset = 78602, Length = 230, Attributes = 0
Repair Log Operations: StreamSegmentAppendOperation: SequenceNumber = 48, SegmentId = 11, Offset = 78832, Length = 114, Attributes = 0
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 49, SegmentId = 11, Attributes = 3
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 50, SegmentId = 11, Attributes = 5
Repair Log Operations: UpdateAttributesOperation: SequenceNumber = 51, SegmentId = 11, Attributes = 2
Number of operations read from DurableLog: 51
```

8) Finally, we repeat the same process with no edits. Which means that the Backup Log and the Repair Log will be the same. After doing this, and given that Original Log 23 is completely fine, we should expect Segment Container 23 to recover, as the next log from the Segment Store demonstrates:
```
2021-11-19 10:53:36,611 5151134 [core-29] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovery started.
2021-11-19 10:53:36,611 5151134 [core-29] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[23]: Enter RecoveryMode.
2021-11-19 10:53:36,612 5151135 [core-29] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[23] Enter RecoveryMode.
2021-11-19 10:53:36,652 5151175 [core-29] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23]: Starting recovery from Sequence Number 1 (skipped 0 Operations and 0 Data Frames).
2021-11-19 10:53:36,652 5151175 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering MetadataCheckpointOperation: SequenceNumber = 1, Length = 999.
2021-11-19 10:53:36,654 5151177 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 2, SegmentId = 11, Attributes = 3.
2021-11-19 10:53:36,654 5151177 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 3, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,654 5151177 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StorageMetadataCheckpointOperation: SequenceNumber = 4, Length = 85.
2021-11-19 10:53:36,655 5151178 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentTruncateOperation: SequenceNumber = 5, SegmentId = 2, Offset = 9438.
2021-11-19 10:53:36,656 5151179 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 6, SegmentId = 2, Offset = 9438, Length = 134, Attributes = 0.
2021-11-19 10:53:36,656 5151179 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 7, SegmentId = 11, Offset = 72245, Length = 699, Attributes = 0.
2021-11-19 10:53:36,656 5151179 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 8, SegmentId = 11, Offset = 72944, Length = 70, Attributes = 0.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 9, SegmentId = 11, Attributes = 4.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 10, SegmentId = 11, Offset = 73014, Length = 230, Attributes = 0.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 11, SegmentId = 11, Offset = 73244, Length = 114, Attributes = 0.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 12, SegmentId = 11, Attributes = 3.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 13, SegmentId = 11, Attributes = 5.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 14, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 15, SegmentId = 2, Offset = 9572, Length = 134, Attributes = 0.
2021-11-19 10:53:36,657 5151180 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 16, SegmentId = 11, Offset = 73358, Length = 751, Attributes = 0.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 17, SegmentId = 11, Offset = 74109, Length = 0, Attributes = 0.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 18, SegmentId = 11, Attributes = 6.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 19, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentTruncateOperation: SequenceNumber = 20, SegmentId = 2, Offset = 9572.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 21, SegmentId = 11, Offset = 74109, Length = 550, Attributes = 0.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 22, SegmentId = 11, Offset = 74659, Length = 70, Attributes = 0.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 23, SegmentId = 11, Attributes = 3.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 24, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 25, SegmentId = 1, Attributes = 2.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 26, SegmentId = 2, Offset = 9706, Length = 92, Attributes = 0.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 27, SegmentId = 2, Offset = 9798, Length = 96, Attributes = 0.
2021-11-19 10:53:36,658 5151181 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentTruncateOperation: SequenceNumber = 28, SegmentId = 2, Offset = 9706.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 29, SegmentId = 11, Offset = 74729, Length = 550, Attributes = 0.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 30, SegmentId = 11, Offset = 75279, Length = 70, Attributes = 0.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 31, SegmentId = 11, Attributes = 3.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 32, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentTruncateOperation: SequenceNumber = 33, SegmentId = 2, Offset = 9894.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 34, SegmentId = 2, Offset = 9894, Length = 134, Attributes = 0.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 35, SegmentId = 11, Offset = 75349, Length = 699, Attributes = 0.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 36, SegmentId = 11, Offset = 76048, Length = 70, Attributes = 0.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 37, SegmentId = 11, Attributes = 4.
2021-11-19 10:53:36,659 5151182 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 38, SegmentId = 11, Offset = 76118, Length = 230, Attributes = 0.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 39, SegmentId = 11, Offset = 76348, Length = 114, Attributes = 0.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 40, SegmentId = 11, Attributes = 3.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 41, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 42, SegmentId = 11, Attributes = 5.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 43, SegmentId = 2, Offset = 10028, Length = 134, Attributes = 0.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 44, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 45, SegmentId = 11, Offset = 76462, Length = 751, Attributes = 0.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 46, SegmentId = 11, Offset = 77213, Length = 0, Attributes = 0.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 47, SegmentId = 11, Attributes = 6.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentTruncateOperation: SequenceNumber = 48, SegmentId = 2, Offset = 10028.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 49, SegmentId = 11, Offset = 77213, Length = 550, Attributes = 0.
2021-11-19 10:53:36,660 5151183 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 50, SegmentId = 11, Offset = 77763, Length = 70, Attributes = 0.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 51, SegmentId = 11, Attributes = 3.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 52, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentTruncateOperation: SequenceNumber = 53, SegmentId = 2, Offset = 10162.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 54, SegmentId = 2, Offset = 10162, Length = 134, Attributes = 0.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 55, SegmentId = 11, Offset = 77833, Length = 699, Attributes = 0.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 56, SegmentId = 11, Offset = 78532, Length = 70, Attributes = 0.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 57, SegmentId = 11, Attributes = 4.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 58, SegmentId = 11, Offset = 78602, Length = 230, Attributes = 0.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering StreamSegmentAppendOperation: SequenceNumber = 59, SegmentId = 11, Offset = 78832, Length = 114, Attributes = 0.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 60, SegmentId = 11, Attributes = 3.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 61, SegmentId = 11, Attributes = 5.
2021-11-19 10:53:36,661 5151184 [core-29] DEBUG i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovering UpdateAttributesOperation: SequenceNumber = 62, SegmentId = 11, Attributes = 2.
2021-11-19 10:53:36,666 5151189 [core-29] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[23] Recovery completed. Epoch = 2, Items Recovered = 62, Time = 54ms.
2021-11-19 10:53:36,666 5151189 [core-29] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[23]: Exit RecoveryMode.
```
Note that the sequence numbers start from 1, which is one of the evidences that the repair tool has been used.
